okhttp/src/main/java/okhttp3/EventListener.java

```diff
@@ -21,6 +21,31 @@
 import java.util.List;
 import javax.annotation.Nullable;
 
+/**
+ * EventListener for analytic events for an OkHttpClient instance.
+ *
+ * <p>All start/connect/acquire events will eventually receive a matching end/release event,
+ * either successful (non-null parameters), or failed (non-null throwable).  The first common
+ * parameters of each event pair are used to link the event in case of concurrent or repeated
+ * events e.g. dnsStart(call, domainName) -> dnsEnd(call, domainName, inetAddressList, throwable).
+ *
+ * <p>Nesting is as follows
+ * <ul>
+ * <li>call -> (dns -> connect -> secure connect)* -> request events</li>
+ * <li>call -> (connection acquire/release)*</li>
+ * </ul>
+ *
+ * <p>Request events are ordered: requestHeaders -> requestBody -> responseHeaders -> responseBody
+ *
+ * <p>Since connections may be reused, the dns and connect events may not be present for a call,
+ * or may be repeated in case of failure retries, even concurrently in case of happy eyeballs type
```

Comment:
Dig. I hope you don’t mind I might go over this in a follow-up. Unlike code which is pretty objective, documentation is a thing that benefits from lots of copy-editing.